File: src/assetmap.cpp
#include "imfwizard/assetmap.h"
#include "imfwizard/uuid.h"
#include <libxml/tree.h>
#include <spdlog/spdlog.h>
#include <ctime>
#include <stdexcept>

namespace imfwizard {

static const char* AM_NS = "http://www.smpte-ra.org/schemas/429-9/2007/AM";

void generate_assetmap(const AssetMapOptions& opts,
                       const std::filesystem::path& output_dir)
{
    std::filesystem::path am_path = output_dir / "ASSETMAP.xml";

    xmlDocPtr doc = xmlNewDoc(BAD_CAST "1.0");
    xmlNodePtr root = xmlNewNode(nullptr, BAD_CAST "AssetMap");
    xmlDocSetRootElement(doc, root);
    xmlNewNs(root, BAD_CAST AM_NS, nullptr);

    xmlNewChild(root, nullptr, BAD_CAST "Id", BAD_CAST generate_uuid().c_str());
    xmlNewChild(root, nullptr, BAD_CAST "Issuer", BAD_CAST opts.issuer.c_str());

    // IssueDate
    time_t now = time(nullptr);
    struct tm tm_buf;
    gmtime_r(&now, &tm_buf);
    char date_buf[64];
    strftime(date_buf, sizeof(date_buf), "%Y-%m-%dT%H:%M:%S+00:00", &tm_buf);
    xmlNewChild(root, nullptr, BAD_CAST "IssueDate", BAD_CAST date_buf);

    // AssetList
    xmlNodePtr asset_list = xmlNewChild(root, nullptr, BAD_CAST "AssetList", nullptr);
    for (const auto& entry : opts.assets) {
        xmlNodePtr asset = xmlNewChild(asset_list, nullptr, BAD_CAST "Asset", nullptr);
        xmlNewChild(asset, nullptr, BAD_CAST "Id", BAD_CAST ("urn:uuid:" + entry.uuid).c_str());

        xmlNodePtr chunk_list = xmlNewChild(asset, nullptr, BAD_CAST "ChunkList", nullptr);
        xmlNodePtr chunk = xmlNewChild(chunk_list, nullptr, BAD_CAST "Chunk", nullptr);
        xmlNewChild(chunk, nullptr, BAD_CAST "Path",
                    BAD_CAST entry.path.string().c_str());
        xmlNewChild(chunk, nullptr, BAD_CAST "Length",
                    BAD_CAST std::to_string(entry.size).c_str());
    }

    xmlSaveFormatFileEnc(am_path.string().c_str(), doc, "UTF-8", 1);
    xmlFreeDoc(doc);

    spdlog::info("Generated ASSETMAP: {}", am_path.string());
}

} // namespace imfwizard

File: rust/crates/imfwizard-core/src/assetmap.rs
use std::io::Write;
use std::path::Path;

/// Write an ASSETMAP.xml file.
pub fn write_assetmap(path: &Path, pkl_uuid: &str, cpl_uuid: &str) -> std::io::Result<()> {
    let am_uuid = uuid::Uuid::new_v4();
    let mut f = std::fs::File::create(path)?;
    writeln!(f, r#"<?xml version="1.0" encoding="UTF-8"?>"#)?;
    writeln!(
        f,
        r#"<AssetMap xmlns="http://www.smpte-ra.org/schemas/429-9/2007/AM">"#
    )?;
    writeln!(f, "  <Id>urn:uuid:{am_uuid}</Id>")?;
    writeln!(f, "  <Creator>IMF Wizard</Creator>")?;
    writeln!(f, "  <AssetList>")?;

    // PKL reference
    writeln!(f, "    <Asset>")?;
    writeln!(f, "      <Id>urn:uuid:{pkl_uuid}</Id>")?;
    writeln!(f, "      <ChunkList>")?;
    writeln!(f, "        <Chunk>")?;
    writeln!(f, "          <Path>PKL_{pkl_uuid}.xml</Path>")?;
    writeln!(f, "        </Chunk>")?;
    writeln!(f, "      </ChunkList>")?;
    writeln!(f, "    </Asset>")?;

    // CPL reference
    writeln!(f, "    <Asset>")?;
    writeln!(f, "      <Id>urn:uuid:{cpl_uuid}</Id>")?;
    writeln!(f, "      <ChunkList>")?;
    writeln!(f, "        <Chunk>")?;
    writeln!(f, "          <Path>CPL_{cpl_uuid}.xml</Path>")?;
    writeln!(f, "        </Chunk>")?;
    writeln!(f, "      </ChunkList>")?;
    writeln!(f, "    </Asset>")?;

    writeln!(f, "  </AssetList>")?;
    writeln!(f, "</AssetMap>")?;
    Ok(())
}
